sensing/src/main/java/com/google/android/sensing/impl/SensingEngineImpl.kt

```diff
@@ -212,6 +213,56 @@ internal class SensingEngineImpl(
     TODO("Not yet implemented")
   }
 
+  override suspend fun resetFailedUploadRequests() {
+    val failedUploadRequestList: List<UploadRequest>  = database.listUploadRequests(status = RequestStatus.FAILED)
+    // Extract the resourceInfoIds from pending requests for quick lookup
+    val pendingResourceIds = database.listUploadRequests(status = RequestStatus.PENDING).map { it.resourceInfoId }.toSet()
+    // Extract the resourceInfoIds from uploaded requests for quick lookup
+    val uploadedResourceIds = database.listUploadRequests(status = RequestStatus.UPLOADED).map { it.resourceInfoId }.toSet()
+    // Filter failed requests whose resourceInfoId is not in pending and completed requests
+    val failedNotInPendingOrCompletedList = failedUploadRequestList
+      .filter { it.resourceInfoId !in pendingResourceIds && it.resourceInfoId !in uploadedResourceIds }
+
+    if (failedNotInPendingOrCompletedList.isNotEmpty()) {
+      failedNotInPendingOrCompletedList.forEach{ failedNotInPendingOrCompleted ->
```

Comment:
This name is enough here.
```suggestion
      failedNotInPendingOrCompletedList.forEach{ failedRequest ->
```